design/object-bucket/object-bucket.md

```diff
@@ -0,0 +1,164 @@
+
+### Use Cases
+
+**Use Case: Expose an Object Store Endpoint for Bucket Provisioning**
+
+_As a cluster admin, I want to expose an object store to cluster users so that they may begin bucket provisioning via the Kubernetes API._
+
+![admin-actions](./obc-admin.png)
+
+In any order, the cluster admin: 
+
+- creates the ObjectBucketClaim Operator (a Deployment API object)
+- creates a StorageClass with `parameters[serviceName]=objectstore-service` and `parameters[serviceNamespace]=some-namespace`
+  - _Note:_ `provisioner` field is left blank.  Defining the operator/provisioner is unnecessary, and if a value is provided it is silently ignored. 
+
+**Use Case: Provision a Bucket** 
+
+_As a Kubernetes user, I want to leverage the Kubernetes API to create S3 buckets. I expect to get back the bucket connection information in a ConfigMap._
+ 
+![user-actions](./obc-user.png)
+
+1. The Rook Operator detects a new ObjectBucketClaim instance.  
+    1. The operator uses `objectBucketClaim.spec.secretName` to get the S3 access keys secret.  
+    1. The operator uses `objectBucketClaim.spec.storageClassName` to get the `Service` endpoint of the object store.
+1. The operator uses the object store endpoint and access keys for an S3 "make bucket" call.
+1. The operator creates a ConfigMap in the namespace of the ObjectBucketClaim with relevant connection data of the bucket.
+1. An app Pod may then mount the Secret and the ConfigMap to begin accessing the bucket. 
+
+**Use Case:** Delete an Object Bucket
+
+_As a Kubernetes user, I want to delete ObjectBucketClaim instances and cleanup generated API resources._
+
+1. The user deletes the ObjectBucketClaim via `kubectl delete ...`.
+1. The ObjectBucketClaim is marked for deletion and left in the foreground.
+1. The respective ConfigMap is deleted.
+1. The ObjectBucketClaim is garbage collected.
+
+---
+
+## API Specifications
+
+### ObjectBucketClaim CRD
+
+```yaml
+apiVersion: apiextensions.k8s.io/v1beta1 
+kind: CustomResourceDefinition
+metadata:
+  name: objectbucketclaims.rook.io
+spec:
+  group: rook.io
+  version: v1alpha2
+  scope: namespaced
+  names:
+      kind: ObjectBucketClaim
+      listKind: ObjectBucketClaimList
+      singular: objectbucketclaim
+      plural: objectbucketclaims
+```
+
+### ObjectBucketClaim
+
+```yaml
+apiVersion: rook.io/v1alpha2
+kind: ObjectBucketClaim
+metadata:
+  name: my-bucket-1
+  namespace: dev-user
+  labels:
+    rook.io/bucket-provisioner:
+    rook.io/object-bucket-claim: my-bucket-1 [1]
+spec:
+  storageClassName: some-object-store
+  secretName: my-s3-key-pair
+  bucketNamePrefix: prefix [2]
+```
+
+1. Added by the rook operator.
+1. The operator will append a hyphen followed by random characters to the string given here.
+
+### Access Keys Secret  
+  
+```yaml
+apiVersion: v1
+kind: Secret
+metadata:
+  name: my-s3-key-pair
+  namespace: dev-user
+data:
+  accessKeyId: <base64 encoded string>
+  secretAccessKey: <base64 encoded string>
```

Comment:
I'd advocate against the use of AWS secrets.
1st AWS is working on [binding IAM to pods](https://github.com/aws/containers-roadmap/issues/23).
2nd in the meantime, I'm fine with scheduling the pod on nodes that have the right instance profile.

Reply:
> I'd advocate against the use of AWS secrets.
1st AWS is working on binding IAM to pods.

Will this work for object stores beside AWS S3?

> 2nd in the meantime, I'm fine with scheduling the pod on nodes that have the right instance profile.

Does this assume multiple cluster users would use the same instance profile?  If user A and user B work within their own namespaces and require access to an object store, would both of their pods potentially use the same instance profile?